### PR TITLE
[LSP] Fix semantic tokens overlap bug

### PR DESCRIPTION
**Pull request by allisonchou**
I inadvertently introduced a bug in https://github.com/dotnet/roslyn/pull/58238, in which we sometimes return both semantic and syntactic classifications for the same TextSpan. In reality, if there is a semantic classification, it should override the syntactic classification for any matching TextSpan unless the syntactic classification is a modifier (e.g. `static`).

This bug introduced was caught via an assert failure in Razor and is currently breaking semantic tokens on their end.

The existing LSP semantic tokens tests were previously partially incorrect. They have been updated this PR and now have the correct expected results, which should help catch this problem in the future.

### CODE REVIEW

**Comment by allisonchou** [comment]
@jinujoseph for M2 approval

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 1]
```diff
@@ -187,7 +188,10 @@ static SemanticTokensHelpers()
                 // Therefore, we set both optional parameters to false.
                 var spans = await ClassifierHelper.GetClassifiedSpansAsync(
                     document, textSpan, cancellationToken, removeAdditiveSpans: false, fillInClassifiedSpanGaps: false).ConfigureAwait(false);
-                classifiedSpans.AddRange(spans);
+
+                // The spans returned to us may include some empty spans, which we don't care about.
+                var nonEmptySpans = spans.Where(s => !s.TextSpan.IsEmpty);
```
Comment (ryzngard) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:193]: What is the use case for empty spans that are classified?

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs [hunk 2]
```diff
@@ -198,10 +198,18 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
...
-                    : syntaxPartAndSpan;
-                finalParts.Add(part);
+                if (replacementIndex >= 0 && !IsClassifiedAsText(semanticParts[replacementIndex]))
+                {
+                    finalParts.Add(semanticParts[replacementIndex]);
+                }
+                // We might already have a semantic part for the given TextSpan, in
+                // which case we don't want to add the syntactic part unless it's a
+                // modifier.
+                else if (finalParts.Count == 0 || !finalParts[^1].TextSpan.Equals(syntaxPartAndSpan.TextSpan) ||
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:208]: nit: this is hard to read as a single line

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs [hunk 3]
```diff
@@ -198,10 +198,18 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
...
-                    ? semanticParts[replacementIndex]
-                    : syntaxPartAndSpan;
-                finalParts.Add(part);
+                if (replacementIndex >= 0 && !IsClassifiedAsText(semanticParts[replacementIndex]))
+                {
+                    finalParts.Add(semanticParts[replacementIndex]);
+                }
+                // We might already have a semantic part for the given TextSpan, in
+                // which case we don't want to add the syntactic part unless it's a
+                // modifier.
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:207]: nit: this says modifier, but the lookup is on additive type names. Are there additive type names that aren't modifiers? Can we use the same terminology here to be consistent?

**Review by ryzngard — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs (hunk 1 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:193]: I can't seem to reproduce it anymore, but there was a file where somehow an empty span was being produced as part of a classification. Razor breaks with empty spans (via a failed assert), so I thought it'd be good to just include this here as a precaution. Empty spans are also unnecessary to return in the LSP case.

File: src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs (hunk 3 above)
Comment (allisonchou) [reply src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs]: Currently there's only one additive type name (static), but I'll change the comment to use consistent terminology. Thanks for the feedback!

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs (hunk 1 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:193]: This actually seems to be a bug in `AdjustSpans` - for now I filed https://github.com/dotnet/roslyn/issues/58658 to track, but it shouldn't block this PR.

**Pull request merged**